libsolidity/codegen/ir/IRGeneratorForStatements.cpp

```diff
@@ -75,18 +75,25 @@ struct CopyTranslate: public yul::ASTCopier
 			{
 				solAssert(reference.isOffset != reference.isSlot, "");
 
-				pair<u256, unsigned> slot_offset = m_context.storageLocationOfVariable(*varDecl);
-
-				string const value = reference.isSlot ?
-					slot_offset.first.str() :
-					to_string(slot_offset.second);
-
-				return yul::Literal{
-					_identifier.location,
-					yul::LiteralKind::Number,
-					yul::YulString{value},
-					{}
-				};
+				string value;
+				if (varDecl->isStateVariable())
+					value =
+						reference.isSlot ?
+							m_context.storageLocationOfStateVariable(*varDecl).first.str() :
+							to_string(m_context.storageLocationOfStateVariable(*varDecl).second);
+				else
+				{
+					solAssert(varDecl->isLocalVariable(), "");
+					if (reference.isSlot)
+						value = IRVariable{*varDecl}.part("slot").name();
+					else
+						value = varDecl->type()->stackItems().size() == 1 ? "0" : IRVariable{*varDecl}.part("offset").name();
```

Comment:
I think it is safer to check for `varDecl->type()->valueType()`.

Reply:
I think there are cases when we have reference type when stack size is 1. For example struct types. Maybe I can create another helper function in `Type` that is more specific?

Reply:
I meant that only value types have an offset part. Non-value types always have an implicit zero offset.

Reply:
Another way to do it would be to check if `offset` exists.

Reply:
But better to add assertions both ways :)

Reply:
I didn't know that all ref types have always only slot part. Will update now.

Reply:
I added some assertions in every part, for both slot and offset...
Hope it is not overkill now :)

Reply:
Actually, `.part` asserts if there is no particular stack slot.
So, I think your original suggestion is enough.